### ISSUE TITLE
Cannot load custom pipeline modules defined in the main script

### ISSUE THREAD

**Issue by allo-** [huggingface/diffusers#515]
### Describe the bug

`pipeline_util.register_modules` tries to retrieve `__module__` from pipeline modules and crashes for modules defined in the main class because the module `__main__` does not contain a `.`

### Reproduction

Create a pipeline with an own safetychecker class, e.g.:

```
class DummyChecker:
    def __call__(self, clip_input, images):
        images, False
```

create a pipe with that module:

```
pipe = StableDiffusionImg2ImgPipeline(
                                    vae,
                                    text_encoder,
                                    tokenizer,
                                    unet,
                                    scheduler,
                                    DummyChecker(),
                                    feature_extractor)
```

### Logs

```shell
With an added `print(name, module.__module__)`:


vae diffusers.models.vae
text_encoder transformers.models.clip.modeling_clip
tokenizer transformers.models.clip.tokenization_clip
unet diffusers.models.unet_2d_condition
scheduler diffusers.schedulers.scheduling_lms_discrete
safety_checker __main__
(...)
  File "diffusers/src/diffusers/pipeline_utils.py", line 88, in register_modules
    pipeline_dir = module.__module__.split(".")[-2]
IndexError: list index out of range
```
```


### System Info

diffusers git checkout

**Comment by patrickvonplaten** [comment]
Haha, I'm not really sure whether we want to help here given the goal of this implementation. 
A bit more effort into hiding its purpose could have helped ;-)

**Comment by allo-** [comment]
My script is using the Pipelines without safetychecker since weeks, I saw no point in hiding it. I think it's 700 MB for something that only benefits one when one wants to provide a public service. *shrugs*

My current issue is, that I need to reload the model to change between generation/img2img/inpainting and I'd like to re-use the already loaded nets, so one can change the pipeline from a prompt in my script. And when I completely deleted the net in the old pipeline, I cannot pass it to the constructor of the new one. Constructing one to delete it again could also work, but I think when the old pipeline already filled the VRAM I cannot even construct a temporary checker.
I didn't put too much energy into trying to construct something to delete it again, though ;-)

So ... I could just plug it into my source checkout of diffusors, but I think each part of the pipeline will have the problem when someone wants to replace it by an own part, so maybe it is better to improve the utils class.

**Comment by patrickvonplaten** [comment]
Ok, let me make a nice community pipeline for all stable diffusion in one early next week since it's requested a lot! Regarding the safety checker that's another topic IMO.

**Comment by allo-** [comment]
Someone had a UnifiedPipeline, but I saw the comment that it would be easier/cleaner to leave them separated?

The problem here wouldn't really require a unified pipeline. Loading `.from_pretrained` works without problems, it's only the constructor that has the check in there.

Regarding optimization that would require modification of the pipeline: Did you consider moving the text_encoder calls out of `.__call__`?
In my Pipeline I interpolate between seeds and had moved the whole `text_embeddings` out of the method, as they are always calculated the same.

At the moment I use vanilla StableDiffusionPipeline again, as I added the option to change promps on the fly, so I cant't delete text_encoder after createding `text_embeddings` anymore. But for a single prompt there could be potential to save VRAM by first calculating the text_embeddings and then using a pipeline that doesn't need the text_encoder anymore.

One could also consider using more advanced python constructs, e.g. using partical calls.
There could, for example, be a method that takes a prompt, runs the text encoder and returns an object, which gets a seed and then computes the images, with fixed text encoding.

**Comment by Karbo123** [comment]
I have the same problem here, when I want to create my own pipeline which is defined in the training script like this:
```
# this is a part of the training script, not a part of the diffusers source codes
class PNDMPipeline(DiffusionPipeline):
    ....
```
Any workaround for this problem? since I don't want to change the diffusers' source codes.

**Comment by Karbo123** [comment]
Just found a simple solution, but not sure what the side effect is:
```python
class PNDMPipeline(DiffusionPipeline):
    def __init__(self, unet, scheduler):
        super().__init__()
        scheduler = scheduler.set_format("pt")
        self.register_to_config(unet=("__main__", "Model"))
        self.register_to_config(scheduler=("diffusers", "PNDMScheduler"))
        self.unet = unet
        self.scheduler = scheduler
```

**Comment by allo-** [comment]
I guess that can work, as long as you don't need logic from the other pipelines, e.g., the `__call__` method with a base and mask image.

I guess in general there are three steps and three similar use-cases:
- Process the prompt (the result can be used with several seeds)
- Diffuse using the prompt and a seed (you can use different seeds, base images, masks)
- SafetyChecker

And the pipelines split it into three applications, just diffusion, img2img and inpainting.

For efficiency it would be useful to do the text_encoding only once when the pipeline is not re-used for other prompts. The safety checker should probably not be loaded into the VRAM while the pipeline is still running, as it only needs to process the results and could do it when the pipeline is finished.
The call methods can and probably should be unified, so one can, for example, save a completely generated image, paint on it and use inpaiting to improve some parts while using the same seed.

I want to publish a script soon, that allows saving, loading and interpolating seeds in the hope to be able to fine-tune generated images later. Between different seeds one could re-use text_encodings. When changing prompts one needs the text_encodings net again, though.
I wonder if one can move the currently not used networks to cpu without completely unloading them in an efficient way.

**Comment by Karbo123** [comment]
Yeah, my provided solution isn't actually good, because I found that if I want to `pipeline.save_pretrained(...)`, it could cause an error. Maybe I just need to extract the logic into another class using `scheduler` or related utils only, and save checkpoint using `torch.save` or something like that to aviod using pipeline.

**Comment by allo-** [comment]
The problem with pipeline modifications is, that it seems that the project still changes quite a lot in the pipelines. I started with a script that contained a `diffuse` method itself which had roughly the content of the `__call__` method and changed to use the Pipeline to get the new features and bugfixes in the meantime.
I think the next larger change may be that the schedulers are replaced and I suppose this can change quite a lot.

**Comment by patrickvonplaten** [comment]
Sorry a very delayed update here, but the above code now works with `diffusers=0.7.1`:

```python
from diffusers import StableDiffusionPipeline
import torch


class DummyChecker:
    def __call__(self, clip_input, images):
        return images, False


pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, revision="fp16", safety_checker=DummyChecker())
pipe = pipe.to("cuda")

prompt = "a photo of an astronaut riding a horse on mars"
image = pipe(prompt).images[0]
```

should not yield any errors.
